### ISSUE TITLE
Difference between AWS-marketplace and API Gateway usage plans?

### ISSUE THREAD

**Issue by vr00n** [argo-marketplace/ARGO-API#4]
API gateway appears to have a feature called Usage Plans which implies that we can create keys and manage their use within API gateway itself:

See here
https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-api-usage-plans.html

AWS marketplace is the actual place to SELL the API. It does not have the logic to manage keys and usage.

If this is indeed the case, then we need to focus on implementing an example `USAGE PLAN` on API GATEWAY for the API we created to add 2 numbers.

Listing it on AWS marketplace is not needed in this scenario

Please confirm if this is right or not right.

Thanks!

# How to submit an issue?

- [x] Have you checked previously open/closed issues to see if your question has already been asked & addressed? 
  - If there is an open issue similar to your issue. Comment on that open issue with additional detail indicating that you too are having the same issue. This allows the project sponsor to prioritize incoming issues.
- [x] Have you provided a clear description of your issue so that the project sponsor is able to answer all your questions with as few back-n-forths as possible?
- [x] (Please acknowledge by checking this box) If a screenshot or visual helps better communicate some complex idea, have you provided one?. `Draw.io` and `lucidchart` are some excellent and quick drawing tools.
- [x] (Please acknowledge by checking this box) When an issue is resolved, please post the steps to resolution or clearly describe the resolution before marking the issue closed. 
   - This will help future `SOLVERS` read through how your issue was solved and may help them solve their issue quicker.
- [x] Have you been kind, courteous, and considerate in your questions? 
- [x] Have you considered using `Stackedit.io`, a handy tool to quickly write in markdown?

**Comment by ankpatil18** [comment]
@vr00n  Just few queries to make sure I understand the requirement correctly 

 1) Do we need to actually deploy the last lambda and API Gateway and then try USAGE PLAN and create and import API Keys   **i.e. actual deployment and implementation** *(as  you have created  aws account for us)* or just the research part of how to use USAGE PLAN once we have API gateway ready?

 2) Whether AWS Marketplace comes into picture for using USAGE PLAN or not ?

Is there any additional requirement or this is current understanding of requirement ?

**Comment by vr00n** [comment]
1. It would be good to show how the `Usage Plan` feature works end to end by testing calls on the dirt simple API you have built and seeing how calls to this API are tracked through the usage plan.

2. I would say finish 1. first and then we can review AWS marketplace.

The goal for you is to assume a business case where you would need to sell access to this API to me (who has expressed a need for an API to add 2 numbers). How would you go about this?

Thanks!

**Comment by ankpatil18** [comment]
@vr00n  done with testing of aws Lambda through passing information API Gateway ( Last  feature we build).

Now to discuss solely on **Usage Plan** here how the whole process works:
- We need to create a Usage Plan
- Under Usage Plan we need all _API_ and _generated API keys_ 
for example let says we have customer1 
    - we create a usage plan **UsagePlanForCustomer1**
    - we generate API Key for customer1, (we can generate multiple key if the customer1 has lot user using it).
    - Last we need to add all API and Keys to the Usage Plan.
In this way we can create N usage Plans with  X numbers of APIs and and Y keys exposed for each plan

_Last step is _AWS MarketPlace_ is where we need to register this Usage Plan. **AWS MarketPlace is definitely needed,** as it sells the plan as a Software as a Service (SaaS) product through AWS Marketplace. API buyers subscribing to the product offering are billed by AWS Marketplace based on the number of requests made to the usage plan._

Also I have created this in AWS account you have created for us, I have changed the password let me know if you want to access and see it for better understanding, will mail you the credential.

Let me know whether it completes the requirement, or we need more information,so will update the read me.
